definition_validator/validator.cc

```diff
@@ -402,9 +404,108 @@ void validateFinalMethodHelper(const core::GlobalState &gs, const core::ClassOrM
             sym.name(gs) == core::Names::unresolvedAncestors()) {
             continue;
         }
-        if (auto e = gs.beginError(sym.loc(gs), core::errors::Resolver::FinalModuleNonFinalMethod)) {
+        auto defLoc = sym.loc(gs);
+        if (auto e = gs.beginError(defLoc, core::errors::Resolver::FinalModuleNonFinalMethod)) {
             e.setHeader("`{}` was declared as final but its method `{}` was not declared as final",
                         errMsgClass.show(gs), sym.name(gs).show(gs));
+
+            // We don't have a direct link to the signature for the method, so we
+            // heuristically look for the matching signature.  This heuristic
+            // assumes that sigs and method definitions appear on separate lines
+            // and would get tripped up by something like:
+            //
+            // sig {...}; def foo; ...; end
+            // def bar; ... end
+            //
+            // finding the `sig` for `bar` even though the `sig` belongs to `foo`.
+            auto source = defLoc.file().data(gs).source();
+            core::LocOffsets sigLocOffsets;
+
+            // We only care about everything prior to the definition.
+            auto line_end = defLoc.beginPos();
+            auto prev_line_end = source.rfind('\n', line_end);
+            bool startFromDefinition = true;
+            while (prev_line_end != source.npos) {
+                auto line_start = prev_line_end + 1;
+                auto line = source.substr(line_start, line_end - line_start);
+                line = absl::StripAsciiWhitespace(line);
+                // We skip empty lines, contra LSP's documentation finder.  But
+                // we permit empty lines (i.e. whitespace) prior to the symbol
+                // definition.
+                if (line.empty()) {
+                    if (!startFromDefinition) {
+                        break;
+                    }
+                    line_end = prev_line_end;
+                    prev_line_end = source.rfind('\n', line_end - 1);
+                    startFromDefinition = false;
+                    continue;
+                }
```

Comment:
Won't this only skip one empty line? Is there a test for this case below?

Reply:
I changed the comment to reflect reality; I was previously skipping empty lines, but decided against doing that...changed the code but didn't really handle the comment.